SOLR-16974: Documentation and changes entry for global CB (#2069)
apache/solr: solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc

circuit breaker thresholds are reached on some nodes. See the <<shards.tolerant Parameter>> for details.

== Circuit Breaker Configurations

> Circuit breakers can be configured globally for the entire node, or for each collection individually, or a combination. Per-collection circit breakers take precedence over global circuit breakers.

Reviewer: ```suggestion
Circuit breakers can be configured globally for the entire node, or for each collection individually, or a combination. Per-collection circuit breakers are checked before global circuit breakers.
```

Both per-collection and global CBs are checked, right? Hmm, not quite sure how to best convey that though.

Author: What I meant with "take precedence" is that if you have a QUERY CB for CPU configured both globally and for collection1, then it is the one from collection1 that will be used / override the global one.

Author: I committed another re-word of the paragraph: "Per-collection circuit breakers are checked before global circuit breakers, and if there is a conflict, the per-collection circuit breaker takes precedence."

Reviewer: Hmm, perhaps I'm misunderstanding the code then ...

... so `getCombinedMap` handles the scenario where both per-collection and global are present by concatenating the lists and the per-collection ones are placed before the global ones: https://github.com/apache/solr/blob/e740123546c7f20c6728df42d493e11f93dc8c79/solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerRegistry.java#L267-L279

... `checkTripped` checks all the circuit breakers of the request type in question but if a tripping one is found it still continues to check the remaining ones because they could have tripped too: https://github.com/apache/solr/blob/e740123546c7f20c6728df42d493e11f93dc8c79/solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerRegistry.java#L168-L181

So my understanding then is that there is ordering in the checks but no precedence or conflict as such because both per-collection and global circuit breakers are always checked?

Reviewer: Thanks for the example, that helps with thinking through the scenarios :-)

> ... e.g. QUERY MEM 90% globally and QUERY MEM 70% for the collection, then a request to that collection will trip at 70%, not 95, since the "local" value takes precedence. ...

That makes me wonder actually, should we consider what happens if someone (mis)configures

> ... e.g. QUERY MEM 70% globally and QUERY MEM 90% for the collection ...

i.e. in that case it will still trip at 70% and the 90% threshold is either a mistake or misunderstanding, or possibly some sort of transitional addition before the global threshold is removed or something.

Or put another way, it's always the lowest threshold that takes precedence?

Author: > Or put another way, it's always the lowest threshold that takes precedence?

I think most users will choose either configuring globally or per core. I really only see utility (from a protect server from over-heating perspective) in the global config. But if a user chooses to do both, my thinking was that they should be allowed to do it in any fashion they like, which includes shooting themselves in the foot by allowing higher or lower values for a particular collection than the global. There may be reasons for both, and I don't think we should dictate what they can do.

The only environment where I can see it would be beneficial to prevent collections from overriding would be in multi tenant environments where you as cluster owner want to maintain full control. But IMO such requirements should come from those kind of users and it can be provided as a future config option to either disable collection-level CB entirely or fine grained precedence rules of some kind. I feel it is premature to guess those needs in this first feature introduction, so keeping it simple.

Reviewer: How about keeping the existing logic and illustrating in the docs what a global/local combination would typically look like e.g.

```
- ... Per-collection circuit breakers are checked before global circuit breakers, and if there is a conflict, the per-collection circuit breaker takes precedence. ...
- ... Per-collection circuit breakers are checked before global circuit breakers, and if there is a conflict, the per-collection circuit breaker takes precedence. Typically, any per-collection circuit breaker thresholds are set lower than global thresholds. ...
```

Author: Done

> A circuit breaker can register itself to trip for query requests and/or update requests. A user may register multiple circuit breakers of the same type with different thresholds for each request type.

Reviewer: ```suggestion
A circuit breaker can register itself to be checked for query requests and/or update requests. A user may register multiple circuit breakers of the same type with different thresholds for each request type.
```

or

```suggestion
A circuit breaker can register itself for query requests and/or update requests. A user may register multiple circuit breakers of the same type with different thresholds for each request type.
```

To enable and configure the JVM heap usage based circuit breaker, add the following:

.Per collection in `solrconfig.xml`
[source,xml]
----
<circuitBreaker class="org.apache.solr.util.circuitbreaker.MemoryCircuitBreaker">
 <double name="threshold">75</double>
</circuitBreaker>
----

.Global in `solr.in.sh`
[source,bash]
----

> SOLR_CIRCUITBREAKER_QUERY_HEAP=75

Reviewer: ```suggestion
SOLR_CIRCUITBREAKER_QUERY_MEM=75
```
